pkg/disttask/importinto/conflict_handler.go

```diff
@@ -0,0 +1,275 @@
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+const (
+	// (1+2+4+8)*0.1s + (10-4)*1s = 7.5s
+	storeOpMinBackoff  = 100 * time.Millisecond
+	storeOpMaxBackoff  = time.Second
+	storeOpMaxRetryCnt = 10
+)
+
+type conflictKVHandler interface {
+	init() error
+	handle(ctx context.Context, key, val []byte) error
+	close() error
+}
+
+type baseConflictKVHandler struct {
+	tableImporter       *importer.TableImporter
+	store               tidbkv.Storage
+	logger              *zap.Logger
+	kvGroup             string
+	handleConflictRowFn func(ctx context.Context, kvGroup string, handle tidbkv.Handle, row []types.Datum, pairs *kv.Pairs) error
+
+	encoder *importer.TableKVEncoder
+}
+
+func (h *baseConflictKVHandler) init() error {
+	encoder, err := h.tableImporter.GetKVEncoderForDupResolve()
+	if err != nil {
+		return err
+	}
+	h.encoder = encoder
+	return nil
+}
+
+func (*baseConflictKVHandler) handle(_ context.Context, _, _ []byte) error {
+	return nil
+}
+
+func (h *baseConflictKVHandler) close() error {
+	return h.encoder.Close()
+}
+
+// re-encode the row from the handle and value of data KV, and delete all encoded keys.
+// it's possible that part or all of the keys are already deleted.
+func (h *baseConflictKVHandler) encodeAndHandleRow(ctx context.Context,
+	encoder *importer.TableKVEncoder, handle tidbkv.Handle, val []byte) (err error) {
+	tbl := h.tableImporter.Table
+	tblMeta := tbl.Meta()
+	decodedData, _, err := tables.DecodeRawRowData(encoder.SessionCtx,
+		tblMeta, handle, tbl.Cols(), val)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	var autoRowID int64
+	if !tblMeta.HasClusteredIndex() {
+		autoRowID = handle.IntValue()
+	}
+	kvPairs, err := encoder.Encode(decodedData, autoRowID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if h.handleConflictRowFn != nil {
+		err = h.handleConflictRowFn(ctx, h.kvGroup, handle, decodedData, kvPairs)
+	} else {
+		err = h.deleteKeysWithRetry(ctx, kvPairs.Pairs)
+	}
+	kvPairs.Clear()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (h *baseConflictKVHandler) deleteKeysWithRetry(ctx context.Context, pairs []common.KvPair) error {
+	backoffer := backoff.NewExponential(storeOpMinBackoff, 2, storeOpMaxBackoff)
+	return dxfhandle.RunWithRetry(ctx, storeOpMaxRetryCnt, backoffer, h.logger, func(ctx context.Context) (bool, error) {
+		err := h.deleteKeys(ctx, pairs)
+		if err != nil {
+			return common.IsRetryableError(err), err
+		}
+		return true, nil
+	})
+}
+
+func (h *baseConflictKVHandler) deleteKeys(ctx context.Context, pairs []common.KvPair) (err error) {
+	txn, err := h.store.Begin()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer func() {
+		if err == nil {
+			err = txn.Commit(ctx)
+		} else {
+			if rollbackErr := txn.Rollback(); rollbackErr != nil {
+				h.logger.Warn("failed to rollback transaction", zap.Error(rollbackErr))
+			}
+		}
+	}()
+
+	for _, p := range pairs {
+		if err = txn.Delete(p.Key); err != nil {
```

Comment:
Will too many keys cause the transaction to be too large and fail?

Reply:
we only delete keys for a single row, it won't be too large normally, but it might be if the row is reaching the limit of TiDB, say 120M, and there are too many indices. but a normal `insert SQL` will also generate this mount of data

I think it's ok, and will add a comment to describe about this behavior